CMakeOptions.txt

```diff
@@ -1,5 +1,7 @@
 # Features that can be enabled for cmake (see CMakeLists.txt)
 
+option(BUILD_SHARED_LIBS "Build using shared libraries" ON)
```

Comment:
`BUILD_SHARED_LIBS` is a CMake cache variable. I would not change the default value in the CMakeLists.txt itself, just pass `-DBUILD_SHARED_LIBS=ON` when invoking cmake.

Reply:
It is not a default cache variable, it is supposed to be set by projects. And by making it an `option` as opposed to `set`, users can see the option when they use `ccmake` or `cmake-gui`.

From https://cmake.org/cmake/help/latest/variable/BUILD_SHARED_LIBS.html#variable:BUILD_SHARED_LIBS
> This variable is often added to projects as an option()

The line is from this tutorial: https://cmake.org/cmake/help/latest/guide/tutorial/index.html#mixing-static-and-shared-step-9

Reply:
You're right, I'm wrong, it isn't a cache variable. However, it doesn't behave any differently whether you specify it as an option or not (It always keeps its value over different invocations). You're also changing the default behaviour of CMake in that by default, libraries are built as static libraries. This is mostly preference but as a user of many CMake projects, I prefer them to stick to the defaults so they all behave the same (Some projects build both types by default which is even worse to deal with). Static libraries is the default so when building a new project I naturally expect to get static libraries.

I don't mind keeping the option, but I would default to `OFF` if only to stick to CMake's defaults.

Reply:
It looks like we have two use cases with different expecations :)

Public libraries (installed in /usr/lib) are usually installed as shared libraries. Static libraries are not always available, especially on Debian/Ubuntu. The current autotools config in ngtcp2 does build both the shared and static libs.

Your use case presumably embeds ngtcp2 in another project which expects ngtcp2 to stay internal, hence static linking. I added a static library earlier for unittests (`ngtcp2_static`), but this could presumably be used by your project as well. It will probably need a special build option (e.g. `ENABLE_STATIC` or whatever is common for CMake projects).

I'll have a look at the common practices for CMake projects and try to come up with something familiar.